File: C_plus_plus/Swap_Nodes_in_Pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *swapPairs(ListNode *head) {
        ListNode *result = head,*left,*right,*last=NULL;
        while(head&&head->next){
            left = head;
            right = head->next;
            head = head->next->next;
            left->next = right->next;
            right->next = left;
            if(!last){
                result = right;
                last = left;
            }
            else{
                last->next = right;
                last = left;
            }
        }
        return result;
    }
};

File: java/Swap_Nodes_in_Pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
 // one pass
public class Solution {
    public ListNode swapPairs(ListNode head) {
        if (head==null || head.next == null) return head;
        ListNode p = head, pre = null;
        head = head.next;
        while(p != null && p.next != null) {
            ListNode after = p.next;
            p.next = after.next;
            after.next = p;
            if (pre!=null)
                pre.next = after;
            pre = p;
            p = p.next;
        }
        return head;
    }
}
